### PR TITLE
Support NVFlare sequence-level classification fine-tuning

### PR DESCRIPTION
**Pull request by farhadrgh**
### Description
This PR adds support for sequence-level classification fine-tuning using ESM2. And refactors the already existing APIs for token-level classification and sequence-level regression.

### Type of changes

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [x]  New feature (non-breaking change which adds functionality)
- [x]  Refactor
- [x]  Documentation update
- [ ]  Other (please describe):

### CI Pipeline Configuration
Configure CI behavior by applying the relevant labels:

- [SKIP_CI](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#skip_ci) - Skip all continuous integration tests
- [INCLUDE_NOTEBOOKS_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_notebooks_tests) - Execute notebook validation tests in pytest
- [INCLUDE_SLOW_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_slow_tests) - Execute tests labelled as slow in pytest for extensive testing

> [!NOTE]
> By default, the notebooks validation tests are skipped unless explicitly enabled.

### Usage

```python
TODO: Add code snippet
```

### Pre-submit Checklist

 - [x] I have tested these changes locally
 - [x] I have updated the documentation accordingly
 - [x] I have added/updated tests as needed
 - [x] All existing tests pass successfully

### CODE REVIEW

**Comment by farhadrgh** [comment]
This is WIP and the following changes are going to be added:
- decomposing the fine-tuning entry point to allow wrapping the trainer run with `flare.patch`
- refactoring ESM2 fine-tune package to `token-level` and `sequence-level` fine-tuning APIs
- bypass `--scale-lr-layer` for LR multiplier and default to tash head parameters
- update notebooks / documentation

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py [hunk 1]
```diff
@@ -13,6 +13,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
```
Comment (trvachov) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py:14]: link comment

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py [hunk 2]
```diff
@@ -44,6 +44,7 @@ def __init__(
         self,
         sequences: pd.Series,
         labels: pd.Series | None = None,
+        task_type: str | None = None,
```
[the remark this answers is not in the document]
Comment (farhadrgh) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:47]: hm, yes, anything else than 'classification', 'regression' will eventually raise a ValueError when setting the loss class. But I will fix the typing here

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py [hunk 3]
```diff
@@ -148,35 +153,48 @@ class InMemorySingleValueDataset(InMemoryProteinDataset):
     def __init__(
         self,
         sequences: pd.Series,
-        labels: pd.Series | None = None,
+        labels: pd.Series,
+        task_type: str = "regression",
```
Comment (pstjohn) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:157]: Same here, use `Literal`

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py [hunk 4]
```diff
@@ -185,7 +203,8 @@ class InMemoryPerTokenValueDataset(InMemoryProteinDataset):
     def __init__(
         self,
         sequences: pd.Series,
-        labels: pd.Series | None = None,
+        labels: pd.Series,
+        task_type: str = "classification",
```
Comment (pstjohn) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:207]: `Literal`

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/loss.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+                backpropagation and the ReductionT will be passed to the reduce method
+                (which currently only works for logging.).
+        """
+        regression_output = forward_out["regression_output"]
+        targets = batch["labels"].to(dtype=regression_output.dtype)  # [b, 1]
+
+        cp_size = parallel_state.get_context_parallel_world_size()
+        if cp_size == 1:
+            loss = torch.nn.functional.mse_loss(regression_output, targets)
+        else:  # TODO: support CP with masked_token_loss_context_parallel
```
Comment (pstjohn) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/loss.py:58]: If you're going to add a TODO comment we should link a Jira ticket or github issue

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/loss.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        cp_size = parallel_state.get_context_parallel_world_size()
+        if cp_size == 1:
+            losses = torch.nn.functional.cross_entropy(classification_output, targets, reduction="none")
+            # token-level losses may contain NaNs at masked locations. We use masked_select to filter out these NaNs
+            if classification_output.dim() == 3:
+                masked_loss = torch.masked_select(losses, loss_mask)
+                loss = masked_loss.sum() / loss_mask.sum()
+            else:
+                loss = losses.mean()  # sequence-level single value classification
+        else:  # TODO: support CP with masked_token_loss_context_parallel
```
Comment (pstjohn) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/loss.py:115]: same here on the TODO comment

File: sub-packages/bionemo-esm2/tests/bionemo/esm2/scripts/test_finetune_esm2.py [hunk 7]
```diff
@@ -145,6 +147,60 @@ def test_esm2_finetune_regressor(
         ), f"Conflict in param requires_grad when encoder_frozen={encoder_frozen}"
 
 
+@pytest.mark.needs_gpu
```
Comment (pstjohn) [inline sub-packages/bionemo-esm2/tests/bionemo/esm2/scripts/test_finetune_esm2.py:150]: I don't think we need this decorator, we don't use it

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/loss.py (hunk 5 above)
Comment (farhadrgh) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/loss.py:58]: hm, this was already there, I think we can remove the comment since we don't expect context parallel here

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/loss.py (hunk 6 above)
Comment (farhadrgh) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/loss.py:115]: yes, will remove the comment

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pstjohn) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:47]: Ah, don't have the default be to raise a `ValueError` 😄

**Comment by pstjohn** [comment]
How does this support NVFlare? it looks more like a refactor of how we select regression or classification?

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py [hunk 8]
```diff
@@ -148,35 +153,48 @@ class InMemorySingleValueDataset(InMemoryProteinDataset):
...
         self,
         sequences: pd.Series,
-        labels: pd.Series | None = None,
+        labels: pd.Series,
+        task_type: str = "regression",
         tokenizer: tokenizer.BioNeMoESMTokenizer = tokenizer.get_tokenizer(),
         seed: int = np.random.SeedSequence().entropy,  # type: ignore
     ):
-        """Initializes a dataset for single-value regression fine-tuning.
+        """Initializes a dataset for single-value fine-tuning.
```
Comment (sichu2023) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:161]: Why a little confused with "single-value" here. Thought it is a base class for both whole-sequence and per-token level classification/regression. Would you mind to clarify it?

**Comment by farhadrgh** [comment]
> How does this support NVFlare? it looks more like a refactor of how we select regression or classification?

They want a sequence-level classification model/finetuning. We only had sequence-level regression and token-=level classification

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py (hunk 8 above)
Comment (farhadrgh) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:161]: yeah single value is expecting one value (regression/classification) for each sequence. But InMemoryPerTokenValueDataset is expecting the labels to be a sequences with the same length as the protein sequences.

**Review by sichu2023 — CHANGES_REQUESTED** [review]
Approve once we have resolved the FLIP v.s. HF head discussion. Maybe we should pull @jstjohn in.

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/sequence_model.py [hunk 9]
```diff
@@ -167,9 +123,17 @@ class ESM2FineTuneSeqConfig(
     # that has this new head and want to keep using these weights, please drop this next line or set to []
     initial_ckpt_skip_keys_with_these_prefixes: List[str] = field(default_factory=lambda: ["regression_head"])
 
+    task_type: str = "regression"
```
[the remark this answers is not in the document]
Comment (farhadrgh) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/sequence_model.py:126]: oh nice catch

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py [hunk 10]
```diff
@@ -197,14 +216,20 @@ def __init__(
...
+        self.task_type = task_type
+        if not task_type == "classification":
+            raise ValueError(f"{task_type} task type is not supported with {self.__class__.__name__}")
+
         label_tokenizer = Label2IDTokenizer()
-        self.label_tokenizer = label_tokenizer.build_vocab("CHE")
+        self.label_tokenizer = label_tokenizer.build_vocab(self.labels.values)
         self.label_cls_eos_id = MLM_LOSS_IGNORE_INDEX
 
     def transform_label(self, label: str) -> Tensor:
```
[the remark this answers is not in the document]
Comment (farhadrgh) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:235]: yes, it could be any string sequences

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py (hunk 8 above)
Comment (sichu2023) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:161]: I see. So the parent class is actually for whole-sequence task.

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/token_model.py [hunk 11]
```diff
@@ -34,63 +33,12 @@
...
-
-        Returns:
-            A tensor that is the mean of the losses. (used for logging).
-        """
-        losses = torch.stack([loss["avg"] for loss in losses_reduced_per_micro_batch])
-        return losses.mean()
-
-
 class MegatronConvNetHead(MegatronModule):
     """A convolutional neural network class for residue-level classification."""
```
[the remark this answers is not in the document]
Comment (farhadrgh) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/token_model.py:43]: yeah, this PR is mainly to support NVFlare use cases. But given the demand for other task heads we should generalize the choice of task head as well.

File: sub-packages/bionemo-esm2/tests/bionemo/esm2/scripts/test_finetune_esm2.py [hunk 12]
```diff
@@ -163,14 +219,14 @@ def mock_parser_args():
         "1",
         "--num-nodes",
         "1",
-        "--min-seq-length",
-        "512",
         "--max-seq-length",
         "1024",
         "--result-dir",
         str(Path("./results")),
```
[the remark this answers is not in the document]
Comment (farhadrgh) [reply sub-packages/bionemo-esm2/tests/bionemo/esm2/scripts/test_finetune_esm2.py:222]: yeah, this test shouldn't create any artifact

**Review by sichu2023 — APPROVED** [review]

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py [hunk 13]
```diff
@@ -71,13 +74,15 @@ def __init__(
...
         csv_path: str | os.PathLike,
+        task_type: Literal["classification", "regression", None] = None,
         tokenizer: tokenizer.BioNeMoESMTokenizer = tokenizer.get_tokenizer(),
         ignore_labels: bool = False,
     ):
         """Class method to create a ProteinDataset instance from a CSV file.
 
         Args:
             csv_path: path to CSV file containing sequences and optionally labels column.
+            task_type (str, optional): Fine-tuning task type. Defaults to None.
```
Comment (pstjohn) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:85]: What does `None` mean though? Doesn't that just cause a failure? Why not just make the default one of classification / regression?

Comment (pstjohn) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:85]: If this is a required argument, then don't provide a default, just make it required

**Review by pstjohn — APPROVED** [review]
LGTM, just left a couple more comments

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py (hunk 13 above)
Comment (farhadrgh) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:85]: `None` is used when we do inference and there is no label in the dataset

Comment (farhadrgh) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:85]: We use `InMemoryProteinDataset` in inference and the other two datasets for finetuning, depending on the label types (single-value, per-token-value)
